### ISSUE TITLE
node-mysql2 does not work with Node.JS 18.12.1 LTS

### ISSUE THREAD

**Issue by ericvgit** [sidorares/node-mysql2#1668]
First, thank you a lot for developing mysql2.
I had various node-mysql2 code blocks to connect to MySQL 8 on Ubuntu 20.04 LTS, which worked with Node.js 16 (LTS). 
None of those code blocks work with Node.js 18 LTS. I isolated the problem to the simple fact that it could not connect to MySQL. I switched to Node.js 16 (LTS), and everything worked again. 
This is very concerning. After investing a lot in moving from PHP to Node.js, now, I have no reliable means for connecting to MySQL when I use an LTS version of Node.js. And this occurred after being forced to move to mysql2 from mysql after installing a newer version of MySQL (version 8). 
It is also concerning because I don't see this problem being reported so far (or not abundantly, so I could find it by Googling or in StackOverflow. And this may be the result of a lack of interest in working with MySQL by the Node.js community and the bias toward using relational databases (e.g., MongoDB).
I even considered Rust to write my connection module, but the fun ends when you have to import a MySQL module similar to mysql2. Eventually, when I considered pure C, I found the same, plus the exciting part: you have to free a memory allocated by the module (but where). 
Bottom line mysql2 has become as essential as express.js, I assume, for those who depend on MySQL. 

Best regards,
Eric V

**Comment by sidorares** [comment]
Hey @ericvgit I added node 18 to the build matrix and tests seem to be running fine
Can you elaborate a bit more what is failing in your setup ( that works with node v16 )? Any errors or other log messages or example code you can share?

**Comment by ericvgit** [comment]
Hello @sidorares
Thank you for developing mysql2 and looking into this problem:
I have two methods to connect to MySQL via mysql2, which both work fine with Node 16 and not with Node 18.12.1 LTS. Also, I use a real web server NGINX/Ubuntu 20.04 LTS and MySQL 8

### **Method 1:**

**_Module mysqlDB.js_**

```
const mysql = require('mysql2');

function DBconnect() {
    const db = mysql.createConnection({
        host: 'localhost',
        user: 'xxx',
        password: 'xxx',
        database: 'xxx'
    })

    // Connect

    db.connect((err) => {
        if(err) {
            //throw err
            console.log("MySQL Error in DBconnect()")
        } else {
            console.log('MySql Connected ...')
        }        
    })
    return db
}

module.exports = DBconnect
```

**_mysql_test1.js_**

```
const DBconnect = require('./control/mysqlDB.js')
db = DBconnect()
let sql = 'SELECT * FROM `sessions` WHERE `name` = ? '
db.query(sql, ['xxx'], function (error, result) {        
    if(error) {
        console.log("MySQL Error", error)
    } else {
        console.log(result[0])
    }
})
db.end()
```

This works fine with Node.js 16. The errors running with Node.js 18 LTS:

```
[nodemon] watching path(s): *.*
[nodemon] watching extensions: js,mjs,json
[nodemon] starting `node mysql_test1.js`
MySQL Error Error: connect ECONNREFUSED ::1:3306
    at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1300:16) {
  errno: -111,
  code: 'ECONNREFUSED',
  syscall: 'connect',
  address: '::1',
  port: 3306,
  fatal: true
}
MySQL Error in DBconnect()
[nodemon] clean exit - waiting for changes before restart

```

**### Method 2:**

**_Module mysqlDB2.js_**

```
const mysql = require('mysql2/promise')

const credentials = {
    host: 'localhost',
    user: 'xxx',
    password: 'xxx',
    database: 'xxx'
}

async function getDB(sql, varArray) {
  let ans = null
  try {
    const db = await mysql.createConnection(credentials)
    ans = await db.execute(sql, varArray)
    db.end()
  } catch(error) {
    console.log("Error in getDB(): ",error)
  }
  return ans
}

module.exports = { 
  getDB
}

```

**mysql_test2.js**

```
const {getDB} = require('./control/mysqlDB2.js')

async function main() {

    let sql = 'SELECT * FROM `sessions` WHERE `name` = ? '
    let varArray = ['xxx']   
    let [rows , fields] = await getDB(sql, varArray)
    console.log("Ans: ", rows)    
}

main()

```

This works with Node.js 16. Errors when using Node.js 18 LTS:

```
[nodemon] 2.0.20
[nodemon] to restart at any time, enter `rs`
[nodemon] watching path(s): *.*
[nodemon] watching extensions: js,mjs,json
[nodemon] starting `node mysql_test2.js`
Error in getDB():  Error: connect ECONNREFUSED ::1:3306
    at Object.createConnection (/var/www/example.com/node_modules/mysql2/promise.js:242:31)
    at getDB (/var/www/example.com/control/mysqlDB2.js:13:28)
    at main (/var/www/example.com/mysql_test2.js:7:33)
    at Object.<anonymous> (/var/www/example.com/mysql_test2.js:11:1)
    at Module._compile (node:internal/modules/cjs/loader:1159:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1213:10)
    at Module.load (node:internal/modules/cjs/loader:1037:32)
    at Module._load (node:internal/modules/cjs/loader:878:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:81:12)
    at node:internal/main/run_main_module:23:47 {
  code: 'ECONNREFUSED',
  errno: -111,
  sqlState: undefined
}
/var/www/example.com/mysql_test2.js:7
    let [rows , fields] = await getDB(sql, varArray)
                          ^

TypeError: (intermediate value) is not iterable
    at main (/var/www/example.com/mysql_test2.js:7:27)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)

Node.js v18.12.1
[nodemon] app crashed - waiting for file changes before starting...


```

Best regards,
Eric V

**Comment by ericvgit** [comment]
Also, referring to my previous post: I use the last version of mysql2 (2.3.3)

**Comment by sidorares** [comment]
can't reproduce, no issues with node 18.9.0 and mysql 8.0.31 ( running locally in docker )

**Comment by kossidts** [comment]
Just a suggestion based on the error portion `Error in getDB():  Error: connect ECONNREFUSED ::1:3306`

I guess node 18 resolves localhost as IPv6 instead on IPv4 (based on the server config). So if the DB user is not allowed to connect from any (%) host, you need to set the host explicitly to the right IP address.

**Comment by sidorares** [comment]
@ericvgit can you try to connect to 127.0.0.1 instead of localhost, e.i explicitly use ipv4 address

**Comment by ericvgit** [comment]
Hello @sidorares,
Thanks a lot for your continued attention to solving this problem. 
I can confirm that by changing the host value (in the code blocks that I have posted above) from 'localhost' to '127.0.0.1' (i.e., ipv4 address for localhost), mysql2 v2.3.3 can connect to MySQL 8.0.31 using Node.js v.18.12.1 (LTS) on Ubuntu 20.04.1 LTS (server). Nevertheless, mysql2 v2.3.3 fails if we use 'localhost' for the host value (which works well with Node.js 16 LTS).
Best regards,
Eric

**Comment by Leccho** [comment]
have you fixed this? I'm having the same problem.

**Comment by sidorares** [comment]
@Leccho the fix is on your side ( if mysql server binds to ipv6 only use ipv6 address of your server instead of 'localhost' )

**Comment by Leccho** [comment]
I've tried `localhost` and `127.0.0.1`, they do not work, I have `root@localhost`, `root@127.0.0.1` and `root@::1` in mysql users. Current mysql version is 8.0.31

I've just upgrade to node 18.12.1LTS and npm 9.2.0. The app did not change and it'S using mysql2 2.3.3 and typeorm 0.3.10 for the connection.

**Comment by sidorares** [comment]
`127.0.0.1` is ipv4 address. Can you try to connect using `::1` instead?

**Comment by trasherdk** [comment]
Well, I don't have any `ipv6` address on any interface, and only one `mysqld` listening:
```
tcp6       0      0 :::3306                 :::*                    LISTEN      1387/mysqld         
```
And no `ipv6` in my connect config:
```
DB_HOST="127.0.0.1"
DB_PORT=3306
DB_NAME="test"
DB_USER="testuser"
DB_PASS="testpass"
```
And still `mysql2` works as expected.
```sh
$ node mysql/src/check-conn-id.js
Acquired connection  8139
Acquired connection  8140
{ conn1: 8139, conn2: 8140 }
```
Trying to connect to `::1` fail, even with a entry in `/etc/hosts` mapping to `localhost`

```
$ node -v && npm -v
v18.12.1
9.2.0
```
